### PR TITLE
Add support for opensubitles login

### PR DESCRIPTION
**Pull request by h3llrais3r**
Make it possible to specify your own username and password for downloading from opensubtitles.

### CODE REVIEW

**Comment by Diaoul** [comment]
Please rebase against master

**Comment by h3llrais3r** [comment]
Do you have an account on opensubtitles for subliminal?
I'm trying to create an account but subliminal and python.subliminal are already taken...

**Comment by h3llrais3r** [comment]
Rebased against master, still waiting on opensubtitles account confirmation before I can commit the tests ;)

**Comment by Diaoul** [comment]
python-subliminal / subliminal for testing purposes.
Yes commit the yaml files.

**Comment by h3llrais3r** [comment]
Done

File: docs/user/cli.rst [hunk 1]
```diff
@@ -13,13 +13,15 @@ subliminal
...
-      --addic7ed USERNAME PASSWORD  Addic7ed configuration.
-      --cache-dir DIRECTORY         Path to the cache directory.  [default:
-                                    ~/.config/subliminal]
-      --debug                       Print useful information for debugging subliminal and for
-                                    reporting bugs.
-      --version                     Show the version and exit.
-      --help                        Show this message and exit.
+      --addic7ed USERNAME PASSWORD    Addic7ed configuration.
+      --opensubtitles USERNAME PASSWORD
+                                      Opensubtitles configuration.
```
Comment (Diaoul) [inline docs/user/cli.rst]: `OpenSubtitles configuration.`

File: subliminal/cli.py [hunk 2]
```diff
@@ -107,13 +107,15 @@ def convert(self, value, param, ctx):
 @click.group(context_settings={'max_content_width': 100}, epilog='Suggestions and bug reports are greatly appreciated: '
              'https://github.com/Diaoul/subliminal/')
 @click.option('--addic7ed', type=click.STRING, nargs=2, metavar='USERNAME PASSWORD', help='Addic7ed configuration.')
+@click.option('--opensubtitles', type=click.STRING, nargs=2, metavar='USERNAME PASSWORD',
+              help='Opensubtitles configuration.')
```
Comment (Diaoul) [inline subliminal/cli.py]: `OpenSubtitles configuration.`

File: subliminal/providers/opensubtitles.py [hunk 3]
```diff
@@ -95,13 +95,19 @@ def get_matches(self, video, hearing_impaired=False):
 class OpenSubtitlesProvider(Provider):
     languages = {Language.fromopensubtitles(l) for l in language_converters['opensubtitles'].codes}
 
-    def __init__(self):
+    def __init__(self, username=None, password=None):
         self.server = ServerProxy('https://api.opensubtitles.org/xml-rpc', TimeoutSafeTransport(10))
+        if username and not password or not username and password:
```
Comment (Diaoul) [inline subliminal/providers/opensubtitles.py]: Why not `if not username or not password`?

File: subliminal/providers/opensubtitles.py [hunk 4]
```diff
@@ -95,13 +95,19 @@ def get_matches(self, video, hearing_impaired=False):
...
+            raise ConfigurationError('Username and password must be specified')
+        # None values not allowed for logging in, so replace it by ''
+        self.username = username if username else ''
+        self.password = password if password else ''
```
Comment (Diaoul) [inline subliminal/providers/opensubtitles.py]: `self.x = x or ''` is more pythonic

File: subliminal/providers/opensubtitles.py [hunk 5]
```diff
@@ -95,13 +95,19 @@ def get_matches(self, video, hearing_impaired=False):
...
         self.token = None
 
     def initialize(self):
         logger.info('Logging in')
-        response = checked(self.server.LogIn('', '', 'eng', 'subliminal v%s' % get_version(__version__)))
+        response = checked(
+            self.server.LogIn(self.username, self.password, 'eng', 'subliminal v%s' % get_version(__version__)))
```
Comment (Diaoul) [inline subliminal/providers/opensubtitles.py]: To be consistent with other long lines:

```python
checked(self.server.LogIn(self.username, self.password, 'eng',
                          'subliminal v%s' % get_version(__version__)))
```

File: tests/test_opensubtitles.py [hunk 6]
```diff
@@ -6,8 +6,8 @@
 import pytest
 from vcr import VCR
 
-from subliminal.providers.opensubtitles import OpenSubtitlesProvider, OpenSubtitlesSubtitle
-
+from subliminal.exceptions import ConfigurationError
+from subliminal.providers.opensubtitles import OpenSubtitlesProvider, OpenSubtitlesSubtitle, OpenSubtitlesError
```
Comment (Diaoul) [inline tests/test_opensubtitles.py]: from subliminal.providers.opensubtitles import OpenSubtitlesError, OpenSubtitlesProvider, OpenSubtitlesSubtitle

I try to keep imports alphabetically ordered (with classes before functions).

**Comment by Diaoul** [comment]
In addition to my inline comments:
* Remove set `token = None` in terminate
* Generic AuthenticationError instead of OpenSubtitlesError

Thanks

**Comment by h3llrais3r** [comment]
Everything is clear, except the remark for the generic AuthenticationError.
Where do you want this? In the opensubtiles.py or the test_opensubtitles.py?

**Comment by Diaoul** [comment]
Both actually, `checked` must be able to raise `AuthenticationError` when login fails.

**Comment by h3llrais3r** [comment]
But what about the Unauthorized exception? It's an AuthenticationError and a OpenSubtitlesError...

**Comment by Diaoul** [comment]
Correct, then catch `Unauthorized` in your tests with `pytest.raises`

**Comment by h3llrais3r** [comment]
Catching of AuthenticationError works also.
Do you prefer the actual Unauthorized exception?

**Comment by Diaoul** [comment]
`Unauthorized`, it's more specific in the provider context. We know it inherits `AuthenticationError` anyway.

**Comment by h3llrais3r** [comment]
Pushed updates.

**Comment by Diaoul** [comment]
Please reopen against develop branch

**Pull request closed without merging**